rust/kernel/align.rs

```diff
@@ -0,0 +1,216 @@
+    (val + mask) & !(mask)
+}
+
+/// Return the alignment of `alignment` bytes of `val`.
+///
+/// `alignment` must be a power of 2, and this function doesn't verify the
+/// requirement. It is the user's responsibility to take charge of this.
+///
+/// This function is const and can be used in const context.
+#[allow(dead_code)]
+#[inline]
+pub const fn align(val: usize, alignment: usize) -> usize {
+    let mask_val: usize = alignment - 1;
+    mask(val, mask_val)
+}
+
+/// Return the alignment downward of `alignment` bytes of `val`.
+///
+/// `alignment` must be a power of 2, and this function doesn't verify the
+/// requirement. It is the user's responsibility to take charge of this.
+///
+/// This function is const and can be used in const context.
+#[allow(dead_code)]
+#[inline]
+pub const fn align_down(val: usize, alignment: usize) -> usize {
+    let val: usize = val - (alignment - 1);
+    align(val, alignment)
+}
+
+/// Test if `val` is `alignment` bytes aligned.
+///
+/// `alignment` must be a power of 2, and this function doesn't verify the
+/// requirement. It is the user's responsibility to take charge of this.
+///
+/// This function is const and can be used in const context.
+#[allow(dead_code)]
+#[inline]
+pub const fn is_aligned(val: usize, alignment: usize) -> bool {
+    (val & (alignment - 1)) == 0
+}
+
+/// An interface for dealing with alignment.
+trait Align {
+    /// The type of alignment value. This type should be integer.
+    /// TODO: Bound this type to integer only.
+    type Alignment;
+
+    /// Alignment by mask
+    fn mask(&self, mask: Self::Alignment) -> Self;
+
+    /// Alignment by bytes.
+    fn align(&self, alignment: Self::Alignment) -> Self;
+
+    /// Alignment downward by bytes.
+    fn align_down(&self, alignment: Self::Alignment) -> Self;
+
+    /// Test if Aligned.
+    fn is_aligned(&self, alignment: Self::Alignment) -> bool;
+}
+
+/// A Helper macro for implementing `Align` trait for primitive integer types.
+macro_rules! impl_align_for_integer {
+    ($type: ty) => {
+        impl Align for $type {
+            type Alignment = Self;
+
+            #[inline]
+            fn mask(&self, mask: Self) -> Self {
+                (self + mask) & !(mask)
+            }
+
+            #[inline]
+            fn align(&self, alignment: Self) -> Self {
+                let mask = alignment - 1;
+                self.mask(mask)
+            }
+
+            #[inline]
+            fn align_down(&self, alignment: Self) -> Self {
+                let val: Self = self - (alignment - 1);
+                val.align(alignment)
+            }
+
+            #[inline]
+            fn is_aligned(&self, alignment: Self) -> bool {
+                (self & (alignment - 1)) == 0
+            }
+        }
+    };
+}
+
+impl_align_for_integer!(u8);
+impl_align_for_integer!(u16);
+impl_align_for_integer!(u32);
+impl_align_for_integer!(u64);
+impl_align_for_integer!(u128);
```

Comment:
Should we implement this for `u128` and `i128`? I wonder if `u128` and `i128` take advantage of FPU, and if it is, we have to forbid to use this.

Reply:
128bit ints should not be used. Many operations on them require compiler-builtins support. All compiler-builtins intrinsics are stubbed out with panicking implementations in rust-for-linux.

Reply:
I removed 128 support.